proxy-libs/heliossqs/heliossqs.go

```diff
@@ -28,27 +34,115 @@ const ServiceAPIVersion = origin_sqs.ServiceAPIVersion
 
 type Client = origin_sqs.Client
 
-func New(options Options,optFns ...func(*Options) ) (*Client) {
-	otelaws.AppendMiddlewares(&options.APIOptions)
-	return origin_sqs.New(options,optFns...)
- }
+type sqsMessageCarrier struct {
+	messageAttrs map[string]types.MessageAttributeValue
+}
+
+func (c sqsMessageCarrier) Get(key string) string {
+	if c.messageAttrs == nil {
+		return ""
+	}
+
+	for attrKey, val := range c.messageAttrs {
+		if attrKey == key {
+			return *val.StringValue
+		}
+	}
+
+	return ""
+}
+
+func (c sqsMessageCarrier) Set(key, val string) {
+	dataType := "String"
+	c.messageAttrs[key] = types.MessageAttributeValue{DataType: &dataType, StringValue: &val}
+}
+
+func (c sqsMessageCarrier) Keys() []string {
+	result := []string{}
+	for key := range c.messageAttrs {
+		result = append(result, key)
+	}
+
+	return result
+}
+
+const maxSqsAttributesCount = 10
+
+func attributeSetter(ctx context.Context, ii middleware.InitializeInput) []attribute.KeyValue {
+	result := []attribute.KeyValue{}
+	switch castParams := ii.Parameters.(type) {
+	case *origin_sqs.SendMessageInput:
+		{
+			attrs := castParams.MessageAttributes
+			if attrs == nil {
+				attrs = map[string]types.MessageAttributeValue{}
+				castParams.MessageAttributes = attrs
+			}
+
+			if len(attrs) < maxSqsAttributesCount {
+				carrier := sqsMessageCarrier{attrs}
+				otel.GetTextMapPropagator().Inject(ctx, carrier)
+			}
+
+			result = append(result, attribute.KeyValue{Key: "messaging.payload", Value: attribute.StringValue(*castParams.MessageBody)})
+		}
+	case *origin_sqs.SendMessageBatchInput:
```

Comment:
I don't understand the other 2 cases here - what's the purpose of them?

Reply:
We want to instrument 3 different types of sqs operations:

1. Send a single message (we want to inject context and collect data)
2. Send multiple message (same as #1)
3. We are received a message we need to make sure that the `traceparent` attribute is collected, as it isn't done by default)

Reply:
why you aren't collecting data in case #2?

Reply:
Done